File: Samples/MainWindow.cs
using System;
using Xwt;
using Xwt.Drawing;

namespace Samples
{
	public class MainWindow: Window
	{
		TreeView samplesTree;
		TreeStore store;
		Image icon;
		VBox sampleBox;
		Label title;
		Widget currentSample;
		
		DataField<string> nameCol = new DataField<string> ();
		DataField<Sample> widgetCol = new DataField<Sample> ();
		DataField<Image> iconCol = new DataField<Image> ();
		
		StatusIcon statusIcon;
		
		public MainWindow ()
		{
			Title = "Xwt Demo Application";
			Width = 500;
			Height = 400;

			try {
				statusIcon = Application.CreateStatusIcon ();
				statusIcon.Menu = new Menu ();
				statusIcon.Menu.Items.Add (new MenuItem ("Test"));
				statusIcon.Image = Image.FromResource (GetType (), "package.png");
			} catch {
				Console.WriteLine ("Status icon could not be shown");
			}
			
			Menu menu = new Menu ();
			
			var file = new MenuItem ("File");
			file.SubMenu = new Menu ();
			file.SubMenu.Items.Add (new MenuItem ("Open"));
			file.SubMenu.Items.Add (new MenuItem ("New"));
			MenuItem mi = new MenuItem ("Close");
			mi.Clicked += delegate {
				Application.Exit();
			};
			file.SubMenu.Items.Add (mi);
			menu.Items.Add (file);
			
			var edit = new MenuItem ("Edit");
			edit.SubMenu = new Menu ();
			edit.SubMenu.Items.Add (new MenuItem ("Copy"));
			edit.SubMenu.Items.Add (new MenuItem ("Cut"));
			edit.SubMenu.Items.Add (new MenuItem ("Paste"));
			menu.Items.Add (edit);
			
			MainMenu = menu;
			
			
			HPaned box = new HPaned ();
			
			icon = Image.FromResource (typeof(App), "class.png");
			
			store = new TreeStore (nameCol, iconCol, widgetCol);
			samplesTree = new TreeView ();
			samplesTree.Columns.Add ("Name", iconCol, nameCol);
			
			AddSample (null, "Boxes", typeof(Boxes));
			AddSample (null, "Buttons", typeof(ButtonSample));
			AddSample (null, "CheckBox", typeof(Checkboxes));
			AddSample (null, "Clipboard", typeof(ClipboardSample));
			AddSample (null, "ColorSelector", typeof(ColorSelectorSample));
			AddSample (null, "ComboBox", typeof(ComboBoxes));
//			AddSample (null, "Designer", typeof(Designer));
			AddSample (null, "Drag & Drop", typeof(DragDrop));
			
			var n = AddSample (null, "Drawing", null);
			AddSample (n, "Canvas with Widget (Linear)", typeof (CanvasWithWidget_Linear));
			AddSample (n, "Canvas with Widget (Radial)", typeof (CanvasWithWidget_Radial));
			AddSample (n, "Chart", typeof (ChartSample));
			AddSample (n, "Colors", typeof(ColorsSample));
			AddSample (n, "Figures", typeof(DrawingFigures));
			AddSample (n, "Transformations", typeof(DrawingTransforms));
			AddSample (n, "Images and Patterns", typeof(DrawingPatternsAndImages));
			AddSample (n, "Text", typeof(DrawingText));
			AddSample (n, "Partial Images", typeof (PartialImages));
			AddSample (n, "Custom Drawn Image", typeof (ImageScaling));

			AddSample (null, "Expander", typeof (ExpanderSample));
			AddSample (null, "Progress bars", typeof(ProgressBarSample));
			AddSample (null, "Frames", typeof(Frames));
			AddSample (null, "Images", typeof(Images));
			AddSample (null, "Labels", typeof(Labels));
			AddSample (null, "ListBox", typeof(ListBoxSample));
			AddSample (null, "LinkLabels", typeof(LinkLabels));
			AddSample (null, "ListView", typeof(ListView1));
			AddSample (null, "Markdown", typeof (MarkDownSample));
			AddSample (null, "Menu", typeof(MenuSamples));
			AddSample (null, "Notebook", typeof(NotebookSample));
			AddSample (null, "Paneds", typeof(PanedViews));
			AddSample (null, "Popover", typeof(PopoverSample));
			AddSample (null, "ReliefFrame", typeof (ReliefFrameSample));
			AddSample (null, "Screens", typeof (ScreensSample));
			AddSample (null, "Scroll View", typeof(ScrollWindowSample));
			AddSample (null, "Spinners", typeof (Spinners));
			AddSample (null, "Tables", typeof (Tables));
			AddSample (null, "Text Entry", typeof (TextEntries));
			AddSample (null, "Tooltips", typeof(Tooltips));
			AddSample (null, "TreeView", typeof(TreeViews));
			AddSample (null, "WidgetEvents", typeof(WidgetEvents));
			AddSample (null, "Windows", typeof(Windows));
			
			samplesTree.DataSource = store;
			
			box.Panel1.Content = samplesTree;
			
			sampleBox = new VBox ();
			title = new Label ("Sample:");
			sampleBox.PackStart (title, BoxMode.None);
			
			box.Panel2.Content = sampleBox;
			box.Panel2.Resize = true;
			box.Position = 160;
			
			Content = box;
			
			samplesTree.SelectionChanged += HandleSamplesTreeSelectionChanged;

			CloseRequested += HandleCloseRequested;
		}

		void HandleCloseRequested (object sender, CloseRequestedEventArgs args)
		{
			args.Handled = !MessageDialog.Confirm ("Samples will be closed", Command.Ok);
		}
		
		protected override void Dispose (bool disposing)
		{
			base.Dispose (disposing);
			
			if (statusIcon != null) {
				statusIcon.Dispose ();
			}
		}

		void HandleSamplesTreeSelectionChanged (object sender, EventArgs e)
		{
			if (samplesTree.SelectedRow != null) {
				if (currentSample != null)
					sampleBox.Remove (currentSample);
				Sample s = store.GetNavigatorAt (samplesTree.SelectedRow).GetValue (widgetCol);
				if (s.Type != null) {
					if (s.Widget == null)
						s.Widget = (Widget)Activator.CreateInstance (s.Type);
					sampleBox.PackStart (s.Widget, BoxMode.FillAndExpand);
				}
				
//				string txt = System.Xaml.XamlServices.Save (s);
				currentSample = s.Widget;
				Dump (currentSample, 0);
			}
		}
		
		void Dump (IWidgetSurface w, int ind)
		{
			if (w == null)
				return;
			Console.WriteLine (new string (' ', ind * 2) + " " + w.GetType ().Name + " " + w.GetPreferredWidth () + " " + w.GetPreferredHeight ());
			foreach (var c in w.Children)
				Dump (c, ind + 1);
		}
		
		TreePosition AddSample (TreePosition pos, string name, Type sampleType)
		{
			//if (page != null)
			//	page.Margin.SetAll (5);
			return store.AddNode (pos).SetValue (nameCol, name).SetValue (iconCol, icon).SetValue (widgetCol, new Sample (sampleType)).CurrentPosition;
		}
	}
	
	class Sample
	{
		public Sample (Type type)
		{
			Type = type;
		}

		public Type Type;
		public Widget Widget;
	}
}

File: SamplesVB/MainWindow.vb
Imports System
Imports Xwt
Imports Xwt.Drawing

Namespace Samples
    Public Class MainWindow
        Inherits Window

        Private samplesTree As TreeView

        Private store As TreeStore

        Private icon As Image

        Private sampleBox As VBox

        Private title As Label

        Private currentSample As Widget

        Private nameCol As DataField(Of String) = New DataField(Of String)()

        Private widgetCol As DataField(Of Sample) = New DataField(Of Sample)()

        Private iconCol As DataField(Of Image) = New DataField(Of Image)()

        Private statusIcon As StatusIcon

        Public Sub New()
            Try
                Me.statusIcon = Application.CreateStatusIcon()
                Me.statusIcon.Menu = New Menu()
                Me.statusIcon.Menu.Items.Add(New MenuItem("Test"))
                Me.statusIcon.Image = Image.FromResource(MyBase.[GetType](), "package.png")
            Catch ex As Exception
                Console.WriteLine("Status icon could not be shown")
            End Try
            Dim menu As Menu = New Menu()
            Dim file As MenuItem = New MenuItem("File")
            file.SubMenu = New Menu()
            file.SubMenu.Items.Add(New MenuItem("Open"))
            file.SubMenu.Items.Add(New MenuItem("New"))
            Dim mi As MenuItem = New MenuItem("Close")
            AddHandler mi.Clicked, Sub()
                                       Application.[Exit]()
                                   End Sub
            file.SubMenu.Items.Add(mi)
            menu.Items.Add(file)
            Dim edit As MenuItem = New MenuItem("Edit")
            edit.SubMenu = New Menu()
            edit.SubMenu.Items.Add(New MenuItem("Copy"))
            edit.SubMenu.Items.Add(New MenuItem("Cut"))
            edit.SubMenu.Items.Add(New MenuItem("Paste"))
            menu.Items.Add(edit)
            MyBase.MainMenu = menu
            Dim box As HPaned = New HPaned()
            Me.icon = Image.FromResource(GetType(App), "class.png")
            Me.store = New TreeStore({Me.nameCol, Me.iconCol, Me.widgetCol})
            Me.samplesTree = New TreeView()
            Me.samplesTree.Columns.Add("Name", {Me.iconCol, Me.nameCol})
            Me.AddSample(Nothing, "Boxes", GetType(Boxes))
            Me.AddSample(Nothing, "Buttons", GetType(ButtonSample))
            Me.AddSample(Nothing, "CheckBox", GetType(Checkboxes))
            Me.AddSample(Nothing, "Clipboard", GetType(ClipboardSample))
            Me.AddSample(Nothing, "ColorSelector", GetType(ColorSelectorSample))
            Me.AddSample(Nothing, "ComboBox", GetType(ComboBoxes))
            Me.AddSample(Nothing, "Drag & Drop", GetType(DragDrop))
            Dim i As TreePosition = Me.AddSample(Nothing, "Drawing", Nothing)
            Me.AddSample(i, "Canvas with Widget", GetType(CanvasWithWidget))
            ' Me.AddSample(i, "Chart", GetType(ChartSample))
            Me.AddSample(i, "Colors", GetType(ColorsSample))
            Me.AddSample(i, "Figures", GetType(DrawingFigures))
            Me.AddSample(i, "Transformations", GetType(DrawingTransforms))
            Me.AddSample(i, "Images and Patterns", GetType(DrawingPatternsAndImages))
            Me.AddSample(i, "Text", GetType(DrawingText))
            Me.AddSample(i, "Partial Images", GetType(PartialImages))
            Me.AddSample(Nothing, "Progress bars", GetType(ProgressBarSample))
            Me.AddSample(Nothing, "Frames", GetType(Frames))
            Me.AddSample(Nothing, "Images", GetType(Images))
            Me.AddSample(Nothing, "Labels", GetType(Labels))
            Me.AddSample(Nothing, "ListBox", GetType(ListBoxSample))
            Me.AddSample(Nothing, "ListView", GetType(ListView1))
            Me.AddSample(Nothing, "Menu", GetType(MenuSamples))
            Me.AddSample(Nothing, "Notebook", GetType(NotebookSample))
            Me.AddSample(Nothing, "Paneds", GetType(PanedViews))
            Me.AddSample(Nothing, "Scroll View", GetType(ScrollWindowSample))
            Me.AddSample(Nothing, "Tables", GetType(Tables))
            Me.AddSample(Nothing, "Text Entry", GetType(TextEntries))
            Me.AddSample(Nothing, "Tooltips", GetType(Tooltips))
            Me.AddSample(Nothing, "TreeView", GetType(TreeViews))
            Me.AddSample(Nothing, "WidgetEvents", GetType(WidgetEvents))
            Me.AddSample(Nothing, "Windows", GetType(Windows))
            Me.samplesTree.DataSource = Me.store
            box.Panel1.Content = Me.samplesTree
            Me.sampleBox = New VBox()
            Me.title = New Label("Sample:")
            Me.sampleBox.PackStart(Me.title, BoxMode.None)
            box.Panel2.Content = Me.sampleBox
            box.Panel2.Resize = True
            box.Position = 160.0
            MyBase.Content = box
            AddHandler Me.samplesTree.SelectionChanged, AddressOf Me.HandleSamplesTreeSelectionChanged
        End Sub

        Protected Overrides Sub Dispose(disposing As Boolean)
            MyBase.Dispose(disposing)
            If Me.statusIcon IsNot Nothing Then
                Me.statusIcon.Dispose()
            End If
        End Sub

        Private Sub HandleSamplesTreeSelectionChanged(sender As Object, e As EventArgs)
            If Me.samplesTree.SelectedRow IsNot Nothing Then
                If Me.currentSample IsNot Nothing Then
                    Me.sampleBox.Remove(Me.currentSample)
                End If
                Dim s As Sample = Me.store.GetNavigatorAt(Me.samplesTree.SelectedRow).GetValue(Of Sample)(Me.widgetCol)
                If s.Type IsNot Nothing Then
                    If s.Widget Is Nothing Then
                        s.Widget = CType(Activator.CreateInstance(s.Type), Widget)
                    End If
                    Me.sampleBox.PackStart(s.Widget, BoxMode.FillAndExpand)
                End If
                Me.currentSample = s.Widget
                Me.Dump(Me.currentSample, 0)
            End If
        End Sub

        Private Sub Dump(w As IWidgetSurface, ind As Integer)
            If w IsNot Nothing Then
                Console.WriteLine(New String(" ", ind * 2), " ", w.[GetType]().Name, " ", w.GetPreferredWidth(), " ", w.GetPreferredHeight())
                For Each c As Widget In w.Children
                    Me.Dump(c, ind + 1)
                Next
            End If
        End Sub

        Private Function AddSample(pos As TreePosition, name As String, sampleType As Type) As TreePosition
            Return Me.store.AddNode(pos).SetValue(Of String)(Me.nameCol, name).SetValue(Of Image)(Me.iconCol, Me.icon).SetValue(Of Sample)(Me.widgetCol, New Sample(sampleType)).CurrentPosition
        End Function
    End Class

    Friend Class Sample
        Public Type As Type

        Public Widget As Widget

        Public Sub New(type As Type)
            Me.Type = type
        End Sub
    End Class

End Namespace
